### PR TITLE
 Vamana build improvement and added docs

### PR DESCRIPTION
**Pull request by bkarsin**
Includes several fixes and improvements to Vamana, primarily:
- Edge case and bug fixes for Vamana index build (details below)
- Documentation added for Vamana
- experimental namespace removed
- Reduce device memory usage by splitting reverse edge work into batches

The edge case fix adds padding to all shared memory size and offset calculations so any dataset dimension is supported (tests added that verify this). A bug was also fixed with the L2 distance metric causing incorrect results in some rare cases. 

This PR addresses the most pressing items in #393 and stabilize the index construction sufficiently to remove the experimental namespace.

### CODE REVIEW

File: docs/source/indexes/vamana.rst [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+The 2 hyper-parameters that are most often tuned are `graph_degree` and `visited_size`. The time needed to create a graph increases dramatically when increasing `graph_degree`, in particular. However, larger graphs may be needed to achieve very high recall search, especially for large datasets.
+
+Memory footprint
+----------------
+
+Vamana builds a graph that is stored in device memory. However, in order to serialize the index and write it to a file for later use, it must be moved into host memory. If the `include_dataset` parameter is also set, then the dataset must be resident in host memory when calling serialize as well. 
+
+Device memory usage
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (cjnolet) [inline docs/source/indexes/vamana.rst:67]: I'm surprised the docs built without error. I'm sure there's a warning in there about having the underlining for headings be longer the text.

File: docs/source/indexes/vamana.rst [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+The built index represents the graph as fixed degree, storing a total of :math:`graph_degree * n_index_vectors` edges. Graph construction also requires the dataset be in device memory (or it copies it to device during build). In addition, device memory is used during construction to sort and create the reverse edges. Thus, the amount of device memory needed depends on the dataset itself, but it is bounded by a maximum sum of:
+
+- vector dataset: :math:`n_index_vectors * n__dims * sizeof(T)`
```
Comment (cjnolet) [inline docs/source/indexes/vamana.rst:71]: I think you might need to escape all of the _'s in here. Have you checked to see how this formats?

Comment (bkarsin) [reply docs/source/indexes/vamana.rst:71]: Did a quick scan but didn't look at all the formatting details. Thanks for checking, I'll do a pass over the format of the final built docs.

File: cpp/src/neighbors/detail/vamana/vamana_build.cuh [hunk 3]
```diff
@@ -253,6 +237,15 @@ void batched_insert_vamana(
       int total_edges;
       raft::copy(&total_edges, d_total_edges.data_handle(), 1, stream);
```
[the remark this answers is not in the document]
Comment (bkarsin) [reply cpp/src/neighbors/detail/vamana/vamana_build.cuh:238]: I thought that because the next line allocates on the same stream, it should be serialized? 
Will add a stream (kinda prefer not to have rmm calls directly in the code, since everything else uses `raft`).

File: cpp/src/neighbors/detail/vamana/robust_prune.cuh [hunk 4]
```diff
@@ -145,9 +145,11 @@ __global__ void RobustPruneKernel(
   // Dynamic shared memory used for blocksort, temp vector storage, and neighborhood list
   extern __shared__ __align__(alignof(ShmemLayout)) char smem[];
 
-  T* s_coords = reinterpret_cast<T*>(&smem[sort_smem_size]);
-  DistPair<IdxT, accT>* new_nbh_list =
-    reinterpret_cast<DistPair<IdxT, accT>*>(&smem[dim * sizeof(T) + sort_smem_size]);
+  int align_padding = (((dim - 1) / alignof(ShmemLayout)) + 1) * alignof(ShmemLayout) - dim;
```
[the remark this answers is not in the document]
Comment (bkarsin) [reply cpp/src/neighbors/detail/vamana/robust_prune.cuh]: Thanks, done.

File: docs/source/indexes/vamana.rst (hunk 1 above)
Comment (bkarsin) [reply docs/source/indexes/vamana.rst]: Fixed all the DiskANN/vamana docs.

File: docs/source/indexes/vamana.rst (hunk 2 above)
Comment (bkarsin) [reply docs/source/indexes/vamana.rst]: Fixed.

File: cpp/src/neighbors/detail/vamana/vamana_build.cuh (hunk 3 above)
[the remark this answers is not in the document]
Comment (tfeher) [reply cpp/src/neighbors/detail/vamana/vamana_build.cuh:238]: That is a good point, but without going through the implementation details of `mdspan`, I it is not clear to me when the value of the `total_edges` variable is read while we prepare the extents of the matrix. Therefore I prefer the explicit sync, thanks for adding it!

Regarding `rmm`, it is legal to use it in cuvs, but it is also fine to stick to `mdarray`.

**Review by tfeher — APPROVED** [review]
Thanks @bkarsin for the updates, the PR looks good to me!
